File: c++/closure.cpp
#include <iostream>

auto foo(int x) {
  return ([=] (int y) -> int {
    return x + y;
  });
}

int main(int argc, char**argv) {
  auto bar = foo(10);
  std::cout<< bar(20) << std::endl;
  return 0;
}

File: haskell/closure.hs
-- demonstration of a closure in haskell

foo x = bar where
  bar y = x + y

main = 
  -- foo returns a function bar with a lexically scoped variable of x.
  -- this function encloses and preserved the scope of the variable
  -- outside its scope of defintion.
  print ((foo 12) 8)

